File: _dsv/head.py
import argparse
from collections import deque
from ._base import _Base
from . import _utils

class head(_Base):
    ''' output the first lines '''
    parser = argparse.ArgumentParser()
    parser.add_argument('-n', '--lines', type=_utils.regex_arg_type(r'-?\d+'), default='10')

    def __init__(self, opts):
        super().__init__(opts)
        self.lines = int(self.opts.lines.group(0))
        self.ring = deque((), -self.lines) if self.opts.lines.group(0).startswith('-') else None
        self.count = 0

    def on_row(self, row):
        if self.ring is None:
            if self.lines == 0:
                return True

            # print first n lines
            super().on_row(row)
            self.count += 1
            if self.count >= self.lines:
                return True
        else:
            # print except for last n lines
            if self.ring and len(self.ring) >= -self.lines:
                super().on_row(self.ring.popleft())
            self.ring.append(row)

File: src/subcommands/head.rs
use crate::utils::Break;
use anyhow::Result;
use crate::base;
use bstr::BString;
use clap::Parser;

#[derive(Parser)]
#[command(about = "output the first lines")]
pub struct Opts {
    #[arg(short = 'n', long, allow_negative_numbers = true, default_value_t = 10, value_name = "NUM", help = "print the first NUM lines")]
    lines: isize,
}

pub struct Handler {
    ring: Option<std::collections::VecDeque<Vec<BString>>>,
    lines: usize,
    count: usize,
}

impl Handler {
    pub fn new(opts: Opts, _base: &mut base::Base) -> Result<Self> {
        Ok(Self {
            ring: if opts.lines >= 0 { None } else { Some(std::collections::VecDeque::new()) },
            lines: opts.lines.unsigned_abs(),
            count: 0,
        })
    }
}


impl base::Processor for Handler {

    fn on_row(&mut self, base: &mut base::Base, row: Vec<BString>) -> Result<()> {
        if let Some(ring) = self.ring.as_mut() {
            // print except for last n lines
            if ring.len() >= self.lines && let Some(row) = ring.pop_front() {
                base.on_row(row)
            } else {
                ring.push_back(row);
                Ok(())
            }

        } else {
            self.count += 1;
            Break::when(self.count > self.lines)
        }
    }
}

